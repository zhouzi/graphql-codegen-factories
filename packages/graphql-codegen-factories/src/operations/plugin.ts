import { Kind, FragmentDefinitionNode, concatAST, DocumentNode } from "graphql";
import {
  oldVisit,
  PluginFunction,
  Types,
} from "@graphql-codegen/plugin-helpers";
import {
  FactoriesOperationsVisitor,
  FactoriesOperationsVisitorRawConfig,
} from "./FactoriesOperationsVisitor";

export const plugin: PluginFunction<
  FactoriesOperationsVisitorRawConfig,
  Types.ComplexPluginOutput
> = (schema, documents, config) => {
  const allAst = concatAST(
    documents.map(({ document }) => document as DocumentNode)
  );
  const fragments = allAst.definitions.filter(
    (d) => d.kind === Kind.FRAGMENT_DEFINITION
  ) as FragmentDefinitionNode[];
  const allFragments: FragmentDefinitionNode[] = [
    ...fragments,
    // `externalFragments` config is passed by the near-operation-file preset.
    // It is an array of fragments declared outside of the operation file.
    ...(config.externalFragments || []).map(({ node }) => node),
  ];

  const visitor = new FactoriesOperationsVisitor(schema, allFragments, config);
  const content = (
    oldVisit(allAst, { leave: visitor }) as DocumentNode
  ).definitions
    .filter((definition) => typeof definition === "string")
    .join("\n");

  return {
    prepend: visitor.getImports(),
    content,
  };
};
