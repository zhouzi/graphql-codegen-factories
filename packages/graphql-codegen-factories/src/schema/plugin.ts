import { printSchema, parse } from "graphql";
import {
  oldVisit,
  PluginFunction,
  Types,
} from "@graphql-codegen/plugin-helpers";
import { FactoriesBaseVisitorRawConfig } from "../FactoriesBaseVisitor";
import { FactoriesSchemaVisitor } from "./FactoriesSchemaVisitor";

export const plugin: PluginFunction<
  FactoriesBaseVisitorRawConfig,
  Types.ComplexPluginOutput
> = (schema, documents, config) => {
  const printedSchema = printSchema(schema);
  const astNode = parse(printedSchema);

  const visitor = new FactoriesSchemaVisitor(schema, config);
  const content = oldVisit(astNode, { leave: visitor })
    .definitions.filter(Boolean)
    .join("\n");

  return {
    prepend: visitor.getImports(),
    content,
  };
};
