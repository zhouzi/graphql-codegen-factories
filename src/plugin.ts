import { printSchema, parse } from "graphql";
import {
  oldVisit,
  PluginFunction,
  Types,
} from "@graphql-codegen/plugin-helpers";
import {
  FactoriesVisitor,
  FactoriesVisitorRawConfig,
} from "./FactoriesVisitor";

export const plugin: PluginFunction<
  FactoriesVisitorRawConfig,
  Types.ComplexPluginOutput
> = (schema, documents, config) => {
  const visitor = new FactoriesVisitor(schema, config);
  const printedSchema = printSchema(schema);
  const astNode = parse(printedSchema);
  const content = oldVisit(astNode, { leave: visitor })
    .definitions.filter(Boolean)
    .join("\n");

  return {
    prepend: visitor.config.typesPath
      ? [
          `import * as ${visitor.config.namespacedImportName} from '${visitor.config.typesPath}';\n`,
        ]
      : [],
    content,
  };
};
