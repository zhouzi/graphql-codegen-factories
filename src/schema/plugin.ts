import { printSchema, parse } from "graphql";
import {
  oldVisit,
  PluginFunction,
  Types,
} from "@graphql-codegen/plugin-helpers";
import {
  FactoriesSchemaVisitor,
  FactoriesSchemaVisitorRawConfig,
} from "./FactoriesSchemaVisitor";

export const plugin: PluginFunction<
  FactoriesSchemaVisitorRawConfig,
  Types.ComplexPluginOutput
> = (schema, documents, config) => {
  const visitor = new FactoriesSchemaVisitor(schema, config);
  const printedSchema = printSchema(schema);
  const astNode = parse(printedSchema);
  const content = oldVisit(astNode, { leave: visitor })
    .definitions.filter(Boolean)
    .join("\n");

  return {
    prepend: visitor.getImports(),
    content,
  };
};
