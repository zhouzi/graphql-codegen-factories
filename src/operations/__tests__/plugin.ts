import { buildSchema, parse } from "graphql";
import { plugin } from "../plugin";

const schema = buildSchema(/* GraphQL */ `
  type User {
    id: ID!
    username: String!
    preferences: UserPreferences!
    followers: [User!]!
  }

  type UserPreferences {
    email: Boolean
    inApp: Boolean
    sms: Boolean
  }

  type Query {
    users: [User!]!
  }

  type Mutation {
    createUser(username: String): User!
  }
`);

describe("plugin", () => {
  it("should generate factories for operations", async () => {
    const ast = parse(/* GraphQL */ `
      mutation CreateUser($username: String) {
        createUser(username: $username) {
          id
          username
          communications: preferences {
            email
          }
          followers {
            id
          }
        }
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "CreateUser.graphql", document: ast }],
      {}
    );
    expect(output).toMatchSnapshot();
  });

  it("should ignore the namespace for the operations types", async () => {
    const ast = parse(/* GraphQL */ `
      mutation CreateUser($username: String) {
        createUser(username: $username) {
          id
          username
        }
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "CreateUser.graphql", document: ast }],
      { namespacedImportName: "types.ts" }
    );
    expect(output).toMatchSnapshot();
  });

  it("should support fragments", async () => {
    const ast = parse(/* GraphQL */ `
      mutation CreateUser($username: String) {
        createUser(username: $username) {
          ...UserFragment
        }
      }
      fragment UserFragment on User {
        id
        username
        preferences {
          ...UserPreferencesFragment
        }
      }
      fragment UserPreferencesFragment on UserPreferences {
        email
        inApp
        sms
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "CreateUser.graphql", document: ast }],
      {}
    );
    expect(output).toMatchSnapshot();
  });
});
