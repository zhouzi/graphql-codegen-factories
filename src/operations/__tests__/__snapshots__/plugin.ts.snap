// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin should generate factories for operations 1`] = `
Object {
  "content": "export function createCreateUserMutationMock(props: Partial<CreateUserMutation>): CreateUserMutation {
  return {
createUser: {
__typename: 'User',
  id: \\"\\",
  username: \\"\\",
  communications: {
  __typename: 'UserPreferences',
    email: null,

  },
  followers: {
  __typename: 'User',
    id: \\"\\",

  },

},

    ...props,
  };
}
",
  "prepend": Array [],
}
`;

exports[`plugin should ignore the namespace for the operations types 1`] = `
Object {
  "content": "export function createCreateUserMutationMock(props: Partial<CreateUserMutation>): CreateUserMutation {
  return {
createUser: {
__typename: 'User',
  id: \\"\\",
  username: \\"\\",

},

    ...props,
  };
}
",
  "prepend": Array [],
}
`;

exports[`plugin should support fragments 1`] = `
Object {
  "content": "export function createCreateUserMutationMock(props: Partial<CreateUserMutation>): CreateUserMutation {
  return {
createUser: {
__typename: 'User',
  id: \\"\\",
  username: \\"\\",
  preferences: {
  __typename: 'UserPreferences',
    email: null,
    inApp: null,
    sms: null,

  },

},

    ...props,
  };
}


",
  "prepend": Array [],
}
`;
